v3

MrJacz opened the pull request:
## Summary

This PR merges the `exportwrappers` branch into `master`, introducing official and highly flexible wrapper exports for multiple popular Discord libraries.

---

## Key Changes

- Now using `tsup` to compile and build
- Now support typescript, commonjs and esm
- README.md complete rewrite to help users
- Added CONTRIBUTING.md
- The wrappers in the library now export with `lavacord/wrappername` instead of having to import/require using filepath


**NOTE** 
CJS and ESM support has not been fully tested





<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced dedicated wrappers for popular Discord libraries (cloudstorm, detritus, discord.js, eris, oceanic) for easier integration.
  - Added a strongly-typed event system for manager and player events.
  - Added a VERSION constant for library version tracking.
  - Added new GitHub Actions workflows for CI, documentation build, and deployment.

- **Refactor**
  - Major internal restructuring for improved modularity, maintainability, and type safety.
  - Unified and modernized REST API handling with enhanced error reporting.
  - Updated player and node classes with clearer state management and event emission.
  - Simplified and modernized TypeScript configuration and build tooling using tsup.
  - Extensive JSDoc documentation added across core classes.

- **Bug Fixes**
  - Improved error handling and reconnection logic for Lavalink nodes.
  - Fixed inconsistencies in player state updates and event emissions.

- **Documentation**
  - Completely overhauled README with detailed usage instructions, examples, and API references.
  - Added a comprehensive contributing guide.
  - Updated funding and license information.

- **Chores**
  - Migrated build system to tsup for multi-module output.
  - Updated dependencies and raised Node.js version requirement to v20.
  - Added and updated configuration files for ESLint, Prettier, and TypeScript.
  - Removed deprecated GitHub Actions workflows for linting and documentation updates.
  - Simplified .gitignore by removing obsolete patterns.
  - Removed TypeScript compilation workflow in favour of consolidated CI workflow.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MrJacz committed:
Export each wrapper seperate

MrJacz committed:
Wrappers are exported now, also added ESM and CJS support

MrJacz committed:
Update readme

MrJacz committed:
fix(README): codacy badge

MrJacz committed:
add(CONTRIBUTING.md)

MrJacz committed:
Completely rewrite the README

MrJacz committed:
fix double up

MrJacz committed:
Update FUNDING.yml

MrJacz committed:
Update(readme): list me and papi as people to donate to. in future will setup lavacord sponsorships

MrJacz committed:
Merge branch 'exportwrappers' of https://github.com/lavacord/Lavacord into exportwrappers

MrJacz committed:
fix(workflow): tsc workflow still using yarn compile now yarn build

MrJacz committed:
fix ignores in eslint

MrJacz committed:
add typescript-eslint and node version to 18

MrJacz committed:
add skipLibCheck

MrJacz committed:
fix link

MrJacz committed:
lint

MrJacz committed:
remove codacy

MrJacz committed:
logo

MrJacz committed:
trans parent logo

MrJacz committed:
missing .js in package.json

MrJacz committed:
fix(tsup): Move types above require/import to avoid tsup warnings when building

MrJacz committed:
Since we now target TS, ESM, CJS using a tsup plugin to get version from package.json as imports/requires are not the same between esm and cjs

MrJacz committed:
Update version to 3

PapiOphidian commented on README.md:
3rd party lib support is mentioned twice here in different sections. I think they should be put next to each other

PapiOphidian reviewed (changes requested):
lgtm other than README mentioning 3rd party libs twice

MrJacz committed:
Added discord-api-types, cleaned up wrappers, added tsdoc lint plugin

MrJacz committed:
Add prettier

MrJacz committed:
Add more TSDoc comments and format all files

MrJacz committed:
Update homepage url to point to lavacord repo

MrJacz committed:
Add non-null operator, until lavalink-types is updated

MrJacz committed:
Remove eslint-plugin-tsdoc

MrJacz committed:
Docs(LavalinkNode): add more information

MrJacz committed:
lint(cloudstorm): remove line

MrJacz committed:
Make state any

MrJacz committed:
remove usage of node.version as only targetting v4 of lavalink now

MrJacz committed:
Remove non-null operators and update to lavalink-types v3

MrJacz committed:
Change license to MIT License

MrJacz committed:
More readme improvments

MrJacz committed:
Export VERSION in index

MrJacz committed:
Remove unused dev dependencies

MrJacz committed:
Add secure option to LavalinkNode

MrJacz committed:
Improve baseRequest and add /v4/info endpoint

MrJacz committed:
Add secure option, add restURL and socketURL get methods, improve LavalinkNode#connect

MrJacz committed:
refactor: Enhance Manager and Player event handling and type definitions

- Updated Manager class to extend EventEmitter with ManagerEvents for better type safety.
- Removed redundant event listener logic in Manager's destroy method.
- Simplified player switching logic in the Manager class.
- Refactored Player class to extend EventEmitter with PlayerEvents for improved event handling.
- Consolidated player state management and event emission for track playback, pause, and volume changes.
- Introduced new methods in Player for setting volume and filters, enhancing clarity and usability.
- Added new REST API method to retrieve Lavalink node statistics.
- Expanded type definitions for Manager and Player events to cover more scenarios and improve type safety.

MrJacz committed:
fix: Correct indentation in FUNDING.yml

MrJacz committed:
chore: Remove emojis from CONTRIBUTING.md

MrJacz committed:
fix: Update Node.js engine requirement to 20

MrJacz committed:
fix: Add alt attribute to contributors image in README.md

MrJacz committed:
fix: Add listener checks before emitting events in Player class

MrJacz committed:
fix: Change state type from any to unknown in LavalinkNodeOptions interface

MrJacz committed:
chore: Remove "How to Contribute" section from Table of Contents in CONTRIBUTING.md

MrJacz committed:
fix: Update tsconfig to allow emitting files and correct path resolution for lavacord

MrJacz committed:
fix: Simplify send method initialization in Manager class

MrJacz committed:
fix: Refactor Manager constructor to simplify options handling and send method initialization

MrJacz committed:
fix: Update README reference in package.json to include file extension

MrJacz committed:
fix: Update return types of Player methods to use UpdatePlayerResult

MrJacz committed:
fix: Update Node.js version requirement in README to v20 or newer

MrJacz committed:
Merge branch 'master' into exportwrappers

MrJacz committed:
fix: Add default export entries for module exports in package.json

MrJacz committed:
fix: Update play method to set options.noReplace and options.userData to undefined instead of deleting

MrJacz committed:
fix: Refactor play method to destructure options and simplify userData handling

MrJacz committed:
fix: Update stats handling in LavalinkNode to use destructuring for improved clarity

MrJacz committed:
chore: Remove CodeQL workflow configuration file

MrJacz committed:
fix: Change return type of switchChannel method from any to unknown for better type safety

MrJacz committed:
fix: Remove duplicate recommended config from ESLint configuration for clarity

MrJacz committed:
fix: Update dependencies in package.json for improved compatibility and performance

MrJacz committed:
fix: Update README to clarify Lavacord's role as a client for Lavalink

MrJacz committed:
fix: Update TypeScript target version to ES2022 in tsconfig and tsup configuration

MrJacz committed:
Bump yarn.lock

MrJacz committed:
feat: Add CI workflows for linting, building, and documentation generation

MrJacz committed:
fix: Refactor CodeQL and documentation workflows for consistency and clarity

MrJacz committed:
fix: Add missing 'with' parameters for CodeQL analysis step

MrJacz committed:
fix: Update CodeQL configuration to use JavaScript instead of TypeScript

MrJacz committed:
fix: Correct Lavalink Setup link in README and update Prettier configuration for better formatting

MrJacz committed:
update: contributing.md

MrJacz committed:
fix: Remove unnecessary properties from Prettier configuration for cleaner formatting

MrJacz committed:
fix: Update README for improved clarity and organization, including logo size adjustment and corrected documentation links

MrJacz committed:
fix: Ensure send function is defined in Manager options and update connect method to return LavalinkNode array

MrJacz committed:
fix: Update TSDoc comments for clarity and accuracy in LavalinkNode class

MrJacz committed:
fix: Emit playerFilters event in setFilters method for manager listeners

MrJacz committed:
fix: Update ManagerEvents and PlayerEvents interfaces with additional event documentation and filter updates

MrJacz committed:
fix: Correct typos in README and Types.ts documentation for clarity

MrJacz committed:
fix: Update README for clarity and consistency, enhancing feature descriptions and table of contents

MrJacz committed:
fix: Correct spelling and phrasing in README for improved clarity

MrJacz committed:
fix: Update configuration files for consistency and improved settings

MrJacz committed:
fix: Improve clarity in README by refining phrasing and correcting typos

MrJacz committed:
fix: Clean up .gitignore and add prepack in package.json scripts

MrJacz committed:
refactor: Rename user property to userId, Allow send method to be defined either by extended Manager class and overriding send method or providing send method in ManagerOptions

MrJacz committed:
refactor(wrappers): Assign userId property on client ready and override send method for all wrappers plus improved event handling/types

PapiOphidian commented:
Hi. You may attempt to modify the cloudstorm wrapper logic for getting shardID, but please note that JS does very funny things in regards to numbers being used as keys in Objects which is what cloudstorm does. I'd recommend leaving it alone, but if you wanna test it, be my guest! Also, recommend bumping cloudstorm to v14 as that was released not too long ago

MrJacz commented on src/wrappers/cloudstorm.ts:
```diff
@@ -0,0 +1,52 @@
+		});
+	}
+
+	protected override send(packet: GatewayVoiceStateUpdate) {
+		if (!this.client.options.totalShards) return;
+
+		const shardID = Number((BigInt(packet.d.guild_id) >> BigInt(22)) % BigInt(this.client.options.totalShards));
+		const s = Object.entries(this.client.shardManager.shards).find((e) => String(e[0]) === String(shardID))?.[1];
+
+		if (s) s.connector.sendMessage(packet);
+	}
```
the current implementation is correct

MrJacz committed:
update: Yarn to v4, add rollup-type-bunder for module augmentation

MrJacz committed:
refactor: Simplify entry points in tsup.config and add typedoc configuration

MrJacz committed:
refactor: Add module documentation and explicitly type VERSION as string

MrJacz committed:
refactor: Update imports and remove use of enum for websocket ops

MrJacz committed:
refactor(Manager): Change userId property to nullable type for better handling of uninitialized state

MrJacz committed:
refactor(Player): Update imports and improve update method return type to APIPlayer

MrJacz committed:
refactor(Rest): Update imports for consistency and improve module structure

MrJacz committed:
refactor(ManagerEvents, PlayerEvents): Add @event annotations to event properties for clarity

MrJacz committed:
refactor(Manager): Make options parameter optional in constructor

MrJacz committed:
fix(Manager): Improve error message formatting for user ID requirement in connect method

MrJacz commented:
> Hi. You may attempt to modify the cloudstorm wrapper logic for getting shardID, but please note that JS does very funny things in regards to numbers being used as keys in Objects which is what cloudstorm does. I'd recommend leaving it alone, but if you wanna test it, be my guest! Also, recommend bumping cloudstorm to v14 as that was released not too long ago

Cloudstorm is bumped up to v14 and no logic is being changed in wrappers :)
